File: module5/closestCarrot.js
const closestCarrot = (grid, startRow, startCol) => {
  let queue = [{ pos: [startRow, startCol], level: 0 }];
  let coordinates = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];

  while (queue.length) {
    let { pos, level } = queue.shift();
    let [row, col] = pos;

    if (grid[row][col] === 'C') return level;
    grid[row][col] = 'X';

    for (let coor of coordinates) {
      let [dx, dy] = coor;
      let new_x = row + dx;
      let new_y = col + dy;

      if (
        new_x >= 0 &&
        new_y >= 0 &&
        new_x < grid.length &&
        new_y < grid[0].length &&
        grid[new_x][new_y] !== 'X'
      ) {
        queue.push({ pos: [new_x, new_y], level: level + 1 });
      }
    }
  }

  return -1;
};

File: module5/closest_carrot.py
from collections import deque

def closest_carrot(grid, starting_row, starting_col):
  '''
  - "shortest path" to carrot -> use bfs
  - write a function to check if in_bounds
  - need some sort of way to track num step step for path
  '''

  def is_inbounds(row, col):
    row_inbounds = 0 <= row < len(grid)
    col_inbounds = 0 <= col < len(grid[0])
    return row_inbounds and col_inbounds

  queue = deque([(starting_row, starting_col, 0)])
  coordinates = [(-1, 0), (1,0), (0, -1), (0, 1)]

  while queue:
    x, y, num_steps = queue.popleft()
    if grid[x][y] == 'C':
      return num_steps
    grid[x][y] = 'X'

    for dx, dy in coordinates:
      new_x, new_y = x + dx, y + dy
      if is_inbounds(new_x, new_y) and grid[new_x][new_y] != 'X':
        queue.append((new_x, new_y, num_steps + 1))

  return -1